7311: add peertask foundation code

Matilda-Clerke opened the pull request:
## PR description
This PR adds the foundational code for the new peer task system. For details about how this is used, please see the[ spike PR](https://github.com/hyperledger/besu/pull/7602) proving the concept is workable.

pinges commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,157 @@
+      }
+    } while (--triesRemaining > 0
+        && executorResult.getResponseCode() != PeerTaskExecutorResponseCode.SUCCESS
+        && executorResult.getResponseCode() != PeerTaskExecutorResponseCode.PEER_DISCONNECTED
```
If you don't care about which peer is used you would try with another peer ...

Matilda-Clerke replied:
This is in the `executeAgainstPeer()` method, so at this point in the code we do care which peer is used. If the calling code doesn't care which peer a task is run against, it can call the `execute()` method instead, which contains the logic for peer switching.

pinges commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,157 @@
+    } while (--triesRemaining > 0
+        && executorResult.getResponseCode() != PeerTaskExecutorResponseCode.SUCCESS
+        && executorResult.getResponseCode() != PeerTaskExecutorResponseCode.PEER_DISCONNECTED
+        && sleepBetweenRetries(WAIT_TIME_BEFORE_RETRY[triesRemaining]));
```
could be handled by the peer manager

Matilda-Clerke replied:
I'm not sure how exactly. The peer manager doesn't (and shouldn't) know anything about whether a peer is requested for a retry or for a first attemp.

siladu replied:
Why do we need to sleep between retries?

Matilda-Clerke replied:
We don't strictly need to sleep between retries, but if a peer is unable to respond before timing out, it's likely just busy or we're experiencing network issues, so waiting a moment before retrying can allow things a chance to improve.

siladu replied:
My understanding was this is a refactor so I think we should match the existing behaviour. Not sure how the retries, possible it is 1 second?  https://github.com/hyperledger/besu/blob/main/ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/AbstractRetryingPeerTask.java#L154

If we can later improve the peering performance with different timeouts that would be a good experiment to test out once the refactor is stable.

Instead of Thread.sleep, is there a reason not to use ethScheduler.scheduleFutureTask as before?

Matilda-Clerke replied:
#Instead of Thread.sleep, is there a reason not to use ethScheduler.scheduleFutureTask as before?
Mainly because this particular method shouldn't be concerned with scheduling tasks. Its singular concern is to implement task execution with optional retries.

In addition, methods in the EthSchedule class seem to operate on either Runnables or EthTasks, which really doesn't combine well with the entirely different paradigm at play with PeerTasks and the PeerTaskExecutor. A better solution would be to ensure we have similar metrics set up in the PeerTaskExecutor.

siladu replied:
> Mainly because this particular method shouldn't be concerned with scheduling tasks. 
If we use it, EthScheduler would be the one concerned with scheduling tasks.

I think we need to maintain the metrics as they are, that's quite a big loss in functionality otherwise. Would much rather we use EthScheduler since a number of maintainers are trying to centralise thread execution here so we can reuse the metrics and testing support in a consistent way across Besu.

Matilda-Clerke replied:
Alright, I'll see if I can do a bit of rework to utilise EthScheduler to capture some metrics.

pinges commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,157 @@
+    }
+  }
+
+  private static boolean isPeerUnused(
```
should be handled by the peer manager

Matilda-Clerke replied:
These methods are here because they are used in the predicate specifically to request peers for the peer task executor. They may not be applicable to all calls to `peerManager.getPeer`.

They certainly could be moved to PeerManager, but that more tightly couples the peer manager to this one specific use case. Maintaining separation is what gives and maintains the simplicity of this code, as compared with the old system.

pinges commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,157 @@
+    return !usedEthPeers.contains(ethPeer);
+  }
+
+  private static boolean isPeerHeightHighEnough(final EthPeer ethPeer, final long requiredHeight) {
```
should be handled by the peer manager

pinges commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,157 @@
+    return ethPeer.chainState().getEstimatedHeight() >= requiredHeight;
+  }
+
+  private static boolean isPeerProtocolSuitable(final EthPeer ethPeer, final String protocol) {
```
should be handled by the peer manager

pinges commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskRequestSender.java:
```diff
@@ -0,0 +1,55 @@
+import java.util.concurrent.TimeoutException;
+
+public class PeerTaskRequestSender {
+  private static final long DEFAULT_TIMEOUT_MS = 20_000;
```
I think at the moment we have a timeout of 5s

Matilda-Clerke replied:
Thanks, I'll update it to 5s for now. In future, it might be good to make this configurable somehow, so end users can tune it to work best in their environment. But that's definitely out of scope for this work.

pinges commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskRequestSender.java:
```diff
@@ -0,0 +1,55 @@
+        (boolean streamClosed, MessageData message, EthPeer peer) -> {
+          responseMessageDataFuture.complete(message);
+        });
+    return responseMessageDataFuture.get(timeoutMs, TimeUnit.MILLISECONDS);
```
I like that!

pinges reviewed (commented):
Left some comments

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerManager.java:
```diff
@@ -0,0 +1,64 @@
+import org.slf4j.LoggerFactory;
+
+/** "Manages" the EthPeers for the PeerTaskExecutor */
+public class PeerManager {
```
An interface to allow easy swapping of different implementations is a fine, but I'm not sure how we'd implement retries in the peer manager... the peer manager is completely separated from the PeerTasks and has no idea what tasks are being executed and whether they succeed or fail.

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,157 @@
+
+  public <T> PeerTaskExecutorResult<T> execute(final PeerTask<T> peerTask) {
+    PeerTaskExecutorResult<T> executorResult;
+    int triesRemaining =
```
I'm not sure I understand how that would work. The peer manager doesn't have any connection to the tasks being run

siladu commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,157 @@
+    return CompletableFuture.supplyAsync(() -> execute(peerTask));
+  }
+
+  public <T> PeerTaskExecutorResult<T> executeAgainstPeer(
```
private?

Matilda-Clerke replied:
It could be private for our existing uses, but afaik, there are some places that the calling code definitely cares about which peer it's requesting data from (particularly around pivot blocks, I think)

siladu commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,157 @@
+    return executorResult;
+  }
+
+  public <T> CompletableFuture<PeerTaskExecutorResult<T>> executeAsync(final PeerTask<T> peerTask) {
```
unused, better IMO to add it in as needed.

Matilda-Clerke replied:
Fair enough, it's easy to remove it and bring it back later. What I'm finding is that the parallelisation seems to work well when done over a whole set of actions anyway.

siladu commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,157 @@
+    return executorResult;
+  }
+
+  public <T> CompletableFuture<PeerTaskExecutorResult<T>> executeAgainstPeerAsync(
```
unused

siladu commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,157 @@
+
+/** Manages the execution of PeerTasks, respecting their PeerTaskBehavior */
+public class PeerTaskExecutor {
+  private static final long[] WAIT_TIME_BEFORE_RETRY = {0, 20000, 5000};
```
What's the thinking behind these retry timings? Warrants a comment I think.

Matilda-Clerke replied:
The idea was to have a progressive back-off timer on retries. As discussed elsewhere, it would be better to stay consistent with the existing implemention (appears to be a 1 second delay), and make this sort of change later, so I'll change that shortly.

siladu replied:
> progressive back-off timer on retries

0 -> 20 seconds -> 5 seconds? Should it `{0, 5000, 20000};` instead?

but yeh, think we should avoid changing any functionality as much as possible

Matilda-Clerke replied:
Oh, it was using the retry countdown to select it. First retry, the countdown would be 2, so `WAIT_TIME_BEFORE_RETRY[triesRemaining]` would be 5000

siladu commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,157 @@
+      try {
+        peer =
+            peerManager.getPeer(
+                (candidatePeer) ->
+                    isPeerUnused(candidatePeer, usedEthPeers)
+                        && (protocolSpecSupplier.get().isPoS()
+                            || isPeerHeightHighEnough(
+                                candidatePeer, peerTask.getRequiredBlockNumber()))
+                        && isPeerProtocolSuitable(candidatePeer, peerTask.getSubProtocol()));
```
Also struggling to find test coverage for this logic.

Matilda-Clerke replied:
So, the idea here is that the calling code supplies a Predicate<EthPeer> to effectively tell the PeerSelector what kind of peer is needed. I suppose if we move the logic to PeerSelector we can just expose additional getPeer methods for different use cases.

siladu commented on ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskRequestSenderTest.java:
```diff
@@ -0,0 +1,73 @@
+                throw new RuntimeException(e);
+              }
+            });
+    Thread.sleep(500);
```
Why do we need this in the test?

Matilda-Clerke replied:
Because of the nature of RequestSender, the actual test has to be performed on another thread than the test thread. The code below this line must be execute after RequestSender has submitted it's callback method to responseStream.then. Since it doesn't make sense to expose a signal just for testing, we can just add a small delay to ensure the code is executed in the required sequence. Obviously, for production code, this is a dirty hack and proper inter-thread signalling should be used.

siladu commented on ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutorTest.java:
Nice coverage!

siladu commented on ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/peertask/DefaultPeerManagerTest.java:
```diff
@@ -0,0 +1,80 @@
+  }
+
+  @Test
+  public void testGetPeer() throws NoAvailablePeerException {
```
would be nice to test the exceptional case too

Matilda-Clerke replied:
Good call, I'll add another test

siladu reviewed (commented):
Ended up making quite a few comments as I went through.

TL;DR: No major objections. The way peers are selected needs some more consideration IMO as it seems a bit spread out across the classes.

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerManager.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.Optional;
+import java.util.function.Predicate;
+
+/** "Manages" the EthPeers for the PeerTaskExecutor */
+public interface PeerManager {
```
Yep, I agree. PeerSelector more accurately describes the classes responsibility.

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerManager.java:
```diff
@@ -0,0 +1,57 @@
+public interface PeerManager {
+
+  /**
+   * Gets the highest reputation peer matching the supplies filter
```
That's fair. Originally, there was no interface, so the javadoc described the implementation. I'll reword it

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/DefaultPeerManager.java:
```diff
@@ -0,0 +1,64 @@
+
+  // use a synchronized map to ensure the map is never modified by multiple threads at once
+  private final Map<PeerId, EthPeer> ethPeersByPeerId =
+      Collections.synchronizedMap(new HashMap<>());
```
Nice, never use that one before.

jframe commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,157 @@
+    MessageData requestMessageData = peerTask.getRequestMessage();
+    PeerTaskExecutorResult<T> executorResult;
+    int triesRemaining =
+        peerTask.getPeerTaskBehaviors().contains(PeerTaskBehavior.RETRY_WITH_SAME_PEER) ? 3 : 1;
```
Bit unsure about this behaviour. If you specify RETRY_WITH_SAME_PEER and RETRY_WITH_OTHER_PEERS what would happen? It would select a peer and try with that peer up to 3 times and if that fails repeat the process up to 3 times. So it could potentially be trying the request up to 9 times?

Matilda-Clerke replied:
Yes, that's correct

jframe commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,157 @@
+    return !usedEthPeers.contains(ethPeer);
+  }
+
+  private static boolean isPeerHeightHighEnough(final EthPeer ethPeer, final long requiredHeight) {
```
This method doesn't need to be static

jframe commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,157 @@
+    return ethPeer.chainState().getEstimatedHeight() >= requiredHeight;
+  }
+
+  private static boolean isPeerProtocolSuitable(final EthPeer ethPeer, final String protocol) {
```
This method doesn't need to be static

jframe commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,157 @@
+    }
+  }
+
+  private static boolean isPeerUnused(
```
This method doesn't need to be static

jframe commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutorResult.java:
```diff
@@ -0,0 +1,35 @@
+
+import java.util.Optional;
+
+public class PeerTaskExecutorResult<T> {
```
This would be simpler as a record

Matilda-Clerke replied:
It means losing the nice Optional wrapping to guarantee the optional result is always set, but yeah in general it should be a record.

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTask.java:
```diff
@@ -0,0 +1,63 @@
+   *
+   * @return the Collection of behaviors this task is expected to exhibit in the PeetTaskExecutor
+   */
+  Collection<PeerTaskBehavior> getPeerTaskBehaviors();
```
Yep true

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTask.java:
```diff
@@ -0,0 +1,63 @@
+   *
+   * @return the SubProtocol used for this PeerTask
+   */
+  String getSubProtocol();
```
Iirc I looked at using the SubProtocol type and decided against it. I can't remember why, so I'll take another look.

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskBehavior.java:
```diff
@@ -0,0 +1,20 @@
+
+public enum PeerTaskBehavior {
+  RETRY_WITH_SAME_PEER,
+  RETRY_WITH_OTHER_PEERS
```
Non-retry behaviour is just the absence of retry behaviours.

I wanted to leave this as PeerTaskBehavior to allow for other behaviors to be added later. They aren't strictly just for describing desired retry behavior, but there aren't currently any other behaviors.

siladu replied:
I think if the goal is for the code to be more readable, I would go with more specific/descriptive naming and let whoever may (or may not) add more behaviours in later worry about the appropriate name then.

Matilda-Clerke replied:
Alright, that makes sense

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,157 @@
+  public <T> PeerTaskExecutorResult<T> execute(final PeerTask<T> peerTask) {
+    PeerTaskExecutorResult<T> executorResult;
+    int triesRemaining =
+        peerTask.getPeerTaskBehaviors().contains(PeerTaskBehavior.RETRY_WITH_OTHER_PEERS) ? 3 : 1;
```
Added constants for now. I think I'll hold off on making them configurable until I get a better idea of how/when/why different values are used.

siladu reviewed (commented):
Few more comments, some of which were discussed on the call.

Thanks for your descriptive commit messages, really help with the re-review!

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/DefaultPeerSelector.java:
```diff
@@ -0,0 +1,102 @@
+  @Override
+  public EthPeer getPeer(
+      final Collection<EthPeer> usedEthPeers,
+      final long requiredPeerHeight,
+      final SubProtocol requiredSubProtocol)
```
I think exposing a method to supply a `Predicate<EthPeer>` in PeerTask can represent the requirements of the task, plus something similar in the executor to represent the requirements of the executor. Stay tuned and I'll update this soon.

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,153 @@
+        peerTask.getPeerTaskBehaviors().contains(PeerTaskRetryBehavior.RETRY_WITH_OTHER_PEERS)
+            ? RETRIES_WITH_OTHER_PEER
+            : NO_RETRIES;
+    final Collection<EthPeer> usedEthPeers = new HashSet<>();
```
That list represents the peers used for this particular task execution. Other tasks executing on the same peer shouldn't have any affect on each other (except through the normal peer reputation feedback mechanism)

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,153 @@
+
+  public <T> CompletableFuture<PeerTaskExecutorResult<T>> executeAgainstPeerAsync(
+      final PeerTask<T> peerTask, final EthPeer peer) {
+    return CompletableFuture.supplyAsync(() -> executeAgainstPeer(peerTask, peer));
```
Yep, I'll be looking at use EthScheduler for this I think

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,153 @@
+
+  public static final int RETRIES_WITH_SAME_PEER = 3;
+  public static final int RETRIES_WITH_OTHER_PEER = 3;
+  public static final int NO_RETRIES = 1;
```
Reworking the loop should be fine.

I'm not sure I see any benefit to putting the number of retries on the PeerTaskRetryBehavior though, especially considering it enumerates different retry behaviors, and shouldn't really provide implementation details of how those enumerations should be implemented.

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTask.java:
```diff
@@ -0,0 +1,64 @@
+   *
+   * @return the Collection of behaviors this task is expected to exhibit in the PeetTaskExecutor
+   */
+  Collection<PeerTaskRetryBehavior> getPeerTaskBehaviors();
```
IntelliJ refactor must have missed this one. I'll update it

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java:
```diff
@@ -465,6 +468,18 @@ public void setTrailingPeerRequirementsSupplier(
+  @Override
+  public EthPeer getPeer(final Predicate<EthPeer> filter) {
+    return streamBestPeers().filter(filter).findFirst().orElseThrow(NoAvailablePeersException::new);
+  }
```
Ok, I'll update

siladu reviewed (commented):
After discussion and further review, it sounds like this code is bypassing some existing functionality in EthPeers. That might be fine if done intentionally, but I'd like to understand why it is fine to skip it.

For example, I think it is missing an equivalent to `EthPeer.hasAvailableRequestCapacity`.

Also `EthPeer.executePeerRequest` has some logic around actualMinBlockNumber and maintains a synchronized `Collection<PendingPeerRequest> pendingRequests` which also has associated metrics - is there an equivalent to these in the new system? Maybe they are superfluous?

Without a good understanding of the old system, in particular EthPeers, it is hard to understand from the new system what is being bypassed, and whether it is intentional or something that was missed. I will have to defer to you @Matilda-Clerke to point these kind of changes out, or maybe for @pinges to spot them.

Matilda-Clerke commented:
> I think it is missing an equivalent to EthPeer.hasAvailableRequestCapacity
I've added this check in getPeer, but it does mean we can't just reuse the bestPeerMatchingCriteria method as you suggested.

> EthPeer.executePeerRequest
There definitely can be an equivalent in the new system. Alternative question though: Is that metric actually useful to anyone? The only metric I can find is just monitoring the size of the List, i.e. monitoring how many PeerRequests are "in flight".
The logic around minimum chain height, I did have implemented in PeerTaskExecutor when selecting a peer, but when we moved to the PeerTask providing a Predicate to filter EthPeers, it wasn't included. I'll have to find a way to get the required information into the places utilising GetReceiptsFromPeerTask so it can be passed in when the task is created.

We definitely need as many eyes on these changes as we can get. I think all that's really missing at the moment is the in-flight requests metric mentioned above.

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,144 @@
+                  peerTask.getPeerRequirementFilter().test(candidatePeer)
+                      && !usedEthPeers.contains(candidatePeer));
+      if (peer.isEmpty()) {
+        executorResult =
```
Yeah I noticed that. Seems very weird to me that we'd just sit and wait for a peer, but only for 5 seconds. Imo, we should just let the task/pipeline fail so it can be retried later.

pinges replied:
With regard to the 5 second wait for a peer, I think there are cases where you want to do that retry, e.g. if all the matching peers are busy. But I think that could probably be decided by the caller?

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,144 @@
+                      && !usedEthPeers.contains(candidatePeer));
+      if (peer.isEmpty()) {
+        executorResult =
+            new PeerTaskExecutorResult<>(
```
No, disconnecting failed peers shouldn't be done here (or in AbstractRetryingSwitchingPeerTask). It's just not one of the responsibilities of a class which is managing the sending of a request message and parsing of the response.

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,144 @@
+      Optional<EthPeer> peer =
+          peerSelector.getPeer(
+              (candidatePeer) ->
+                  peerTask.getPeerRequirementFilter().test(candidatePeer)
```
Yes, all of a tasks requirements should be included in the getPeerRequirementFilter Predicate, and any additional requirements should be added here by the PeerTaskExecutor

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,144 @@
+  private boolean sleepBetweenRetries() {
+    try {
+      // sleep for 1 second to match implemented wait between retries in AbstractRetryingPeerTask
+      Thread.sleep(1000);
```
It's possible. In practice, this is only called on a worker thread anyway

Matilda-Clerke commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,144 @@
+          MessageData responseMessageData =
+              requestSender.sendRequest(peerTask.getSubProtocol(), requestMessageData, peer);
+
+          result = peerTask.parseResponse(responseMessageData);
```
Currently, it's not. Looks like it's typically just checking that the response Collection/Map/etc is empty, which should be easy to add to PeerTask. In the old code, all it's doing is resetting the retry count when a partial success is detected, so I'm not sure it's really necessary to handle.

Matilda-Clerke replied:
Should be handled appropriately now

siladu commented:
> I think all that's really missing at the moment is the in-flight requests metric mentioned above.

Is an equivalent of this logic needed too?https://github.com/hyperledger/besu/blob/aac7c63b0866d78edccdabace956c32947172090/ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java#L268-L275

Matilda-Clerke commented:
@siladu As that's a requirement of the individual task, I've handled in the EthPeer filter of each PeerTask. You can see an example in https://github.com/hyperledger/besu/pull/7638/files#diff-65888fb86ed73e694cc6171b8f99c9150241acb7edc94a575a0ea83b4b91849dR111-R115

jframe commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,201 @@
+            BesuMetricCategory.PEERS,
+            "PeerTaskExecutor:RequestTime",
+            "Time taken to send a request and receive a response",
+            "className");
```
Using label `taskName` would be better and matches with how label used elsewhere

jframe commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,201 @@
+    requestTimer =
+        metricsSystem.createLabelledTimer(
+            BesuMetricCategory.PEERS,
+            "PeerTaskExecutor:RequestTime",
```
Think `request_time` would be better no need for the PeerTaskExecutor: prefix

jframe commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,201 @@
+    this.peerSelector = peerSelector;
+    this.requestSender = requestSender;
+    requestTimer =
+        metricsSystem.createLabelledTimer(
```
Can we just these metric names to use snakecase so they are consistent with naming we are using elsewhere for metrics

jframe commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,201 @@
+            "PeerTaskExecutor:RequestTime",
+            "Time taken to send a request and receive a response",
+            "className");
+    partialSuccessCounter =
```
For counters the convention we are using generally and there are probably exceptions is to end with the total suffix with snakecase.

So I think this would be better as `partial_success_total`

jframe commented on metrics/core/src/main/java/org/hyperledger/besu/metrics/prometheus/PrometheusGauge.java:
```diff
@@ -47,17 +47,20 @@ public PrometheusGauge(
   }
 
+  @Override
+  public boolean isLabelsObserved(final String... labelValues) {
```
Couldn't see where this new method is used? Is it in the other PR?

Matilda-Clerke replied:
It should be used in PeerTaskExecutor.executeAgainstPeer to add labels for tasks we haven't seen before

pinges commented on ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/peertask/PeerTaskExecutor.java:
```diff
@@ -0,0 +1,201 @@
+          MessageData responseMessageData =
+              requestSender.sendRequest(peerTask.getSubProtocol(), requestMessageData, peer);
+
+          result = peerTask.parseResponse(responseMessageData);
```
It might be a good idea to not do the parsing here. I don't think that would be hard to change. It would make the PeerTaskExecutor almost trivial (no peer handling, no retries). The tasks themselves would have to handle that, but I guess they have to handle the partial success anyways. 
Could we have an AbstractTask that does most of that work?
Maybe something we can discuss tomorrow?

Matilda-Clerke replied:
The whole point of the executor is handle retries and peer switching for a task. The task classes themselves don't actively do anything, they're more like a collection of methods that do small specific actions or supply values, like generating the request message. Changing this would be a pretty substantial change at this stage, and creating an AbstractPeerTask class is starting to sound extremely similar to the old system we're trying to move away from.

pinges commented on plugin-api/build.gradle:
```diff
@@ -71,7 +71,7 @@ Calculated : ${currentHash}
   description = "Checks that the API for the Plugin-API project does not change without deliberate thought"
   files = sourceSets.main.allJava.files
-  knownHash = '4jVaj9yW88nHbX0KmTR3dPQRvj9x8Pvh5E9Ry7KRT6w='
+  knownHash = 'VN2JB2HPpEUDQaDvd7QcMkmmgedasVChfA8tnSf1GHU='
```
what has changed in the plugin API?

Matilda-Clerke replied:
Added a method to LabelledGauge to allow us to check whether a label is observed or not

siladu replied:
Can we add a default method (returns a dummy false value?) to avoid the breaking change please?

Matilda-Clerke replied:
Yep, good idea

pinges reviewed (approved):
LGTM
